File: array/leetcode/977.go
package leetcode

func sortedSquares(nums []int) []int {
	n := len(nums)
	left, right, k := 0, n-1, n-1
	ans := make([]int, n)

	for left <= right {
			if nums[left] * nums[left] < nums[right] * nums[right] {
					ans[k] = nums[right] * nums[right]
					right--
			} else {
					ans[k] = nums[left] * nums[left]
					left++
			}
			k--
	}
	return ans
}

File: array/cpp/977.cpp
#include <vector>
using namespace std;

class Solution
{
public:
  vector<int> sortedSquares(vector<int> &nums)
  {
    int k = nums.size() - 1;
    vector<int> result(nums.size(), 0);
    for (int i = 0, j = nums.size() - 1; i <= j;)
    {
      if (nums[i] * nums[i] < nums[j] * nums[j])
      {
        result[k--] = nums[j] * nums[j];
        j--;
      }
      else
      {
        result[k--] = nums[i] * nums[i];
        i++;
      }
    }
    return result;
  }
};